internal/jobobject/queue.go

```diff
@@ -0,0 +1,96 @@
+package jobobject
+
+import (
+	"errors"
+	"sync"
+)
+
+var (
+	errQueueClosed = errors.New("the queue is closed for reading and writing")
+	errQueueEmpty  = errors.New("the queue is empty")
+)
+
+// messageQueue represents a threadsafe message queue to be used to retrieve or
+// write notifications to. This will be used to hold and retrieve messages for job
+// objects and doesn't have much use in other packages. There should only be one reader
+// and one writer.
+type messageQueue struct {
+	m        sync.Mutex
+	messages []interface{}
+	ready    chan struct{} // Channel to signal that a message is available to read.
+	closed   bool
+}
+
+func newMessageQueue() *messageQueue {
+	return &messageQueue{
+		messages: []interface{}{},
+		ready:    make(chan struct{}),
+	}
+}
+
+func (mq *messageQueue) write(msg interface{}) error {
+	mq.m.Lock()
+	defer mq.m.Unlock()
+
+	if mq.closed {
+		return errQueueClosed
+	}
+	length := len(mq.messages)
+	mq.messages = append(mq.messages, msg)
+	// If the queue was empty, we need to signal that there's now a value ready to read.
+	if length == 0 {
+		// This handles the very first write case. If this is the first write, we could end
+		// up in a spot where we'd signal mq.ready but no one is waiting for a value so we'll block
+		// forever. This way we can add to the queue, so any call to read after will just return the
+		// value and any call to readOrWait should be listening for a new value and we'll hit the
+		// `mq.ready <- struct{}{}` case.
+		select {
+		case mq.ready <- struct{}{}:
+		default:
+		}
```

Comment:
I'm confused here. In the description you say that "we could end up in a spot where we'd signal mq.Ready but no one is waiting.... we'll block forever" but we're using a switch statement when sending ready here so there would be no block in that case since we'd just take the default case. Am I just misreading the comment? If so, could you update to clarify?

Reply:
The comment is describing the reason for the select statement itself which could be more clear. If you take this select out and just leave the `mq.ready <- struct{}{}` statement, if on the very first write to the queue (the length will always be 0, reason for the "handles the very first write case" comment) if we don't have something reading mq.ready, e.g. no one called readOrWait yet, we'd just block forever.

Reply:
Okay, I think something like 

```
// Use a select statement here to prevent an issue where there is no one waiting for the value on the first write to the message queue, causing us to hang forever trying to send the ready message .
```

Reply:
Also this would handle more than just the very first write case. It would handle all cases in which the write queue gets emptied before a given call to `write`.